ffi/diplomat/src/datetime_formatter.rs

```diff
@@ -0,0 +1,332 @@
+        ) -> DiplomatResult<Box<ICU4XGregorianDateTimeFormatter>, ICU4XError> {
+            use icu_provider::serde::AsDeserializingBufferProvider;
+            let provider = provider.0.as_deserializing();
+
+            let locale = locale.to_datalocale();
+
+            let options = length::Bag::from_date_time_style(date_length.into(), time_length.into());
+
+            TypedDateTimeFormatter::try_new_unstable(&provider, &locale, options.into())
+                .map(|dtf| Box::new(ICU4XGregorianDateTimeFormatter(dtf)))
+                .map_err(Into::into)
+                .into()
+        }
+
+        /// Formats a [`ICU4XGregorianDateTime`] to a string.
+        #[diplomat::rust_link(icu::datetime::TypedDateTimeFormatter::format_to_write, FnInStruct)]
+        #[diplomat::rust_link(icu::datetime::TypedDateTimeFormatter::format, FnInStruct, hidden)]
+        #[diplomat::rust_link(
+            icu::datetime::TypedDateTimeFormatter::format_to_string,
+            FnInStruct,
+            hidden
+        )]
+        pub fn format_datetime(
+            &self,
+            value: &ICU4XGregorianDateTime,
+            write: &mut diplomat_runtime::DiplomatWriteable,
+        ) -> DiplomatResult<(), ICU4XError> {
+            #[allow(unused_variables)]
+            let result = self
+                .0
+                .format_to_write(write, &value.0)
+                .map_err(Into::into)
+                .into();
+            write.flush();
+            result
+        }
+    }
+
+    #[diplomat::opaque]
+    /// An ICU4X DateFormatter object capable of formatting a [`ICU4XDate`] as a string,
+    /// using some calendar specified at runtime in the locale.
+    #[diplomat::rust_link(icu::datetime::date_length, Struct)]
+    pub struct ICU4XDateFormatter(pub DateFormatter);
+
+    impl ICU4XDateFormatter {
+        /// Creates a new [`ICU4XDateFormatter`] from locale data.
+        #[diplomat::rust_link(icu::datetime::DateTimeFormatter::try_new_unstable, FnInStruct)]
+        pub fn try_new(
+            provider: &ICU4XDataProvider,
+            locale: &ICU4XLocale,
+            date_length: ICU4XDateLength,
+        ) -> DiplomatResult<Box<ICU4XDateFormatter>, ICU4XError> {
+            use icu_provider::serde::AsDeserializingBufferProvider;
+            let provider = provider.0.as_deserializing();
+
+            let locale = locale.to_datalocale();
+
+            DateFormatter::try_new_unstable(&provider, &locale, date_length.into())
+                .map(|dtf| Box::new(ICU4XDateFormatter(dtf)))
+                .map_err(Into::into)
+                .into()
+        }
+
+        /// Formats a [`ICU4XDate`] to a string.
+        #[diplomat::rust_link(icu::datetime::DateFormatter::format_to_write, FnInStruct)]
+        #[diplomat::rust_link(icu::datetime::DateFormatter::format, FnInStruct, hidden)]
+        #[diplomat::rust_link(icu::datetime::DateFormatter::format_to_string, FnInStruct, hidden)]
+        pub fn format_datetime(
```

Comment:
Nit: Add `format_date`, `format_iso_date`, and `format_iso_datetime`

Reply:
We don't currently support format_date-style methods for DateTimeFormatter in Rust, you need something that's also an IsoTimeInput